crates/sui-core/src/transaction_manager.rs

```diff
@@ -0,0 +1,161 @@
+            tx_ready_certificates,
+        };
+        transaction_manager
+            .enqueue(node_sync_store.all_pending_certs().unwrap())
+            .expect("Initialize TransactionManager with pending certificates failed.");
+        transaction_manager
+    }
+
+    /// Enqueues a certificate into TransactionManager. Once all of its input objects are available
+    /// locally, it will be published.
+    /// TODO: turn this function async, and wait for inflight certificates below threshold.
+    pub(crate) fn enqueue(&mut self, certs: Vec<VerifiedCertificate>) -> SuiResult<()> {
+        for cert in certs {
+            let (_, missing, _) = self
+                .authority_store
+                .get_sequenced_input_objects(cert.digest(), &cert.signed_data.data.input_objects()?)
+                .expect("Are shared object locks set prior to enqueueing certificates?");
+            if missing.is_empty() {
+                self.certificate_ready(cert);
+                continue;
+            }
+            let cert_key = (cert.epoch(), *cert.digest());
+            for obj_key in missing {
+                // TODO: verify the key does not already exist.
+                self.missing_inputs.insert(obj_key, cert_key);
+                self.pending_certificates
+                    .entry(cert_key)
+                    .or_default()
+                    .insert(obj_key);
+            }
+        }
+        self.metrics
+            .transaction_manager_num_missing_objects
+            .set(self.missing_inputs.len() as i64);
+        self.metrics
+            .transaction_manager_num_pending_certificates
+            .set(self.pending_certificates.len() as i64);
+        Ok(())
+    }
+
+    /// Notifies TransactionManager that the given objects have been committed.
+    // TODO: switch TransactionManager to use notify_read() on objects table.
+    pub(crate) fn objects_committed(&mut self, object_keys: Vec<ObjectKey>) {
+        for object_key in object_keys {
+            let cert_key = if let Some(key) = self.missing_inputs.remove(&object_key) {
+                key
+            } else {
+                continue;
+            };
+            let set = self.pending_certificates.entry(cert_key).or_default();
+            set.remove(&object_key);
+            // This certificate has no missing input. It is ready to execute.
+            if set.is_empty() {
+                self.pending_certificates.remove(&cert_key);
+                // NOTE: failing and ignoring the certificate is fine, if it will be retried at a higher level.
+                // Otherwise, this has to crash.
+                let cert = match self.node_sync_store.get_cert(cert_key.0, &cert_key.1) {
+                    Ok(Some(cert)) => cert,
+                    Ok(None) => {
+                        error!(
+                            "Ready certificate not found in the pending table: {:?}",
+                            cert_key
```

Comment:
Please log as `error!(tx_digest = ?cert_key, "Ready certificate not found in the pending table")`